File: firmware/telegram.go
//go:build tinygo && !modemonly

package main

import (
	"bytes"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"runtime"
	"strconv"
	"strings"
	"time"

	"github.com/dimajolkin/sms-telegram/firmware/swtls"
)

type Telegram struct {
	token  string
	client http.Client
	tr     *swtls.Transport
}

type Update struct {
	UpdateID int64      `json:"update_id"`
	Message  *TGMessage `json:"message"`
}

type TGMessage struct {
	MessageID int64  `json:"message_id"`
	Text      string `json:"text"`
	Chat      TGChat `json:"chat"`
}

type TGChat struct {
	ID int64 `json:"id"`
}

func (t *Telegram) apiURL(method string) string {
	return "https://api.telegram.org/bot" + t.token + "/" + method
}

func (t *Telegram) Send(chatID, text string) error {
	form := url.Values{}
	form.Set("chat_id", chatID)
	form.Set("text", text)
	req, err := http.NewRequest(http.MethodPost, t.apiURL("sendMessage"), strings.NewReader(form.Encode()))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("User-Agent", "sms-telegram/1.0")
	body, code, err := t.do(req, 512)
	if err != nil {
		return err
	}
	if code >= 300 {
		return errStr("telegram http " + strconv.Itoa(code) + " " + preview(body))
	}
	return nil
}

func (t *Telegram) GetUpdates(offset int64, timeoutSec int) ([]Update, int64, error) {
	q := url.Values{}
	q.Set("timeout", strconv.Itoa(timeoutSec))
	q.Set("limit", "1")
	if offset > 0 {
		q.Set("offset", strconv.FormatInt(offset, 10))
	}
	u := t.apiURL("getUpdates") + "?" + q.Encode()
	req, err := http.NewRequest(http.MethodGet, u, nil)
	if err != nil {
		return nil, offset, err
	}
	req.Header.Set("User-Agent", "sms-telegram/1.0")
	body, code, err := t.do(req, 4*1024)
	if err != nil {
		return nil, offset, err
	}
	if code >= 300 {
		return nil, offset, errStr("http " + strconv.Itoa(code) + " " + preview(body))
	}
	if len(body) == 0 {
		return nil, offset, errStr("empty body")
	}
	// пустой poll без json.Unmarshal — экономит heap
	if bytes.Contains(body, []byte(`"result":[]`)) {
		return nil, offset, nil
	}
	var parsed struct {
		OK          bool     `json:"ok"`
		Result      []Update `json:"result"`
		Description string   `json:"description"`
	}
	if err := json.Unmarshal(body, &parsed); err != nil {
		return nil, offset, errStr(err.Error() + " " + preview(body))
	}
	if !parsed.OK {
		return nil, offset, errStr("api: " + parsed.Description)
	}
	next := offset
	for _, u := range parsed.Result {
		if u.UpdateID >= next {
			next = u.UpdateID + 1
		}
	}
	return parsed.Result, next, nil
}

func (t *Telegram) Ping() error {
	req, err := http.NewRequest(http.MethodGet, t.apiURL("getMe"), nil)
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", "sms-telegram/1.0")
	body, code, err := t.do(req, 512)
	if err != nil {
		return err
	}
	if code >= 300 {
		return errStr("getMe http " + strconv.Itoa(code) + " " + preview(body))
	}
	if len(body) == 0 || body[0] != '{' {
		return errStr("getMe not json " + preview(body))
	}
	println("tg getMe:", preview(body))
	return nil
}

// SetMyCommands ставит меню «/» из botCommands.
func (t *Telegram) SetMyCommands() error {
	// компактный JSON без encoding/json — меньше heap
	var b strings.Builder
	b.WriteString(`{"commands":[`)
	for i, c := range botCommands {
		if i > 0 {
			b.WriteByte(',')
		}
		b.WriteString(`{"command":"`)
		b.WriteString(c.Cmd)
		b.WriteString(`","description":"`)
		b.WriteString(escapeJSON(c.Desc))
		b.WriteString(`"}`)
	}
	b.WriteString(`]}`)
	payload := b.String()
	req, err := http.NewRequest(http.MethodPost, t.apiURL("setMyCommands"), strings.NewReader(payload))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "sms-telegram/1.0")
	body, code, err := t.do(req, 256)
	if err != nil {
		return err
	}
	if code >= 300 {
		return errStr("setMyCommands http " + strconv.Itoa(code) + " " + preview(body))
	}
	return nil
}

func escapeJSON(s string) string {
	s = strings.ReplaceAll(s, `\`, `\\`)
	s = strings.ReplaceAll(s, `"`, `\"`)
	return s
}

func (t *Telegram) do(req *http.Request, maxBody int) ([]byte, int, error) {
	t.ensureClient()
	resp, err := t.client.Do(req)
	if err != nil {
		t.tr.CloseIdle()
		runtime.GC()
		return nil, 0, err
	}
	body, err := io.ReadAll(io.LimitReader(resp.Body, int64(maxBody)))
	_ = resp.Body.Close()
	// Не держим keep-alive: Conn+bufio держат десятки KB, UI/модем не влезают.
	t.tr.CloseIdle()
	runtime.GC()
	if err != nil {
		return nil, resp.StatusCode, err
	}
	return body, resp.StatusCode, nil
}

func (t *Telegram) ensureClient() {
	if t.tr == nil {
		t.tr = &swtls.Transport{Timeout: 45 * time.Second}
		t.client.Transport = t.tr
		t.client.Timeout = 45 * time.Second
	}
}

func preview(b []byte) string {
	s := string(b)
	s = strings.ReplaceAll(s, "\n", " ")
	s = strings.ReplaceAll(s, "\r", "")
	if len(s) > 80 {
		s = s[:80] + "…"
	}
	return strconv.Quote(s)
}

File: firmware/src/telegram.cpp
#include "telegram.h"
#include "commands.h"
#include <ArduinoJson.h>
#include <HTTPClient.h>
#include <WiFiClientSecure.h>
#include <cctype>
#include <cstdio>
#include <cstring>

void Telegram::begin(const char *token) { token_ = token ? token : ""; }

String Telegram::apiURL(const char *method) const {
  return String("https://api.telegram.org/bot") + token_ + "/" + method;
}

bool Telegram::doRequest(const char *method, const char *url,
                         const char *payload, const char *content_type,
                         String *body_out, int *code_out, int max_body) {
  WiFiClientSecure client;
  client.setInsecure();
  HTTPClient https;
  https.setTimeout(45000);
  if (!https.begin(client, url)) {
    Serial.println("tg begin failed");
    return false;
  }
  https.setUserAgent("sms-telegram/2.0");
  int code;
  if (payload && content_type) {
    https.addHeader("Content-Type", content_type);
    code = https.POST((uint8_t *)payload, strlen(payload));
  } else if (strcmp(method, "POST") == 0) {
    code = https.POST("");
  } else {
    code = https.GET();
  }
  if (code_out) *code_out = code;
  String body = https.getString();
  if ((int)body.length() > max_body) body = body.substring(0, max_body);
  if (body_out) *body_out = body;
  https.end();
  return code > 0;
}

bool Telegram::send(const char *chat_id, const char *text) {
  if (!chat_id || !chat_id[0] || !text) return false;
  String url = apiURL("sendMessage");
  String form = "chat_id=" + String(chat_id) + "&text=";
  // URL-encode minimally
  for (const char *p = text; *p; p++) {
    char c = *p;
    if (c == ' ')
      form += '+';
    else if (isalnum((unsigned char)c) || c == '-' || c == '_' || c == '.' ||
             c == '~')
      form += c;
    else {
      char hex[8];
      snprintf(hex, sizeof(hex), "%%%02X", (unsigned char)c);
      form += hex;
    }
  }
  String body;
  int code = 0;
  if (!doRequest("POST", url.c_str(), form.c_str(),
                 "application/x-www-form-urlencoded", &body, &code, 512))
    return false;
  if (code >= 300) {
    Serial.printf("tg send http %d %s\n", code, body.c_str());
    return false;
  }
  return true;
}

bool Telegram::ping() {
  String url = apiURL("getMe");
  String body;
  int code = 0;
  if (!doRequest("GET", url.c_str(), nullptr, nullptr, &body, &code, 512))
    return false;
  if (code >= 300 || body.length() == 0 || body[0] != '{') {
    Serial.printf("tg getMe fail %d %s\n", code, body.c_str());
    return false;
  }
  Serial.printf("tg getMe: %s\n", body.substring(0, 80).c_str());
  return true;
}

bool Telegram::setMyCommands() {
  String payload = "{\"commands\":[";
  for (int i = 0; i < BOT_COMMANDS_N; i++) {
    if (i) payload += ',';
    payload += "{\"command\":\"";
    payload += BOT_COMMANDS[i].cmd;
    payload += "\",\"description\":\"";
    // escape quotes in desc
    for (const char *p = BOT_COMMANDS[i].desc; *p; p++) {
      if (*p == '"' || *p == '\\') payload += '\\';
      payload += *p;
    }
    payload += "\"}";
  }
  payload += "]}";
  String url = apiURL("setMyCommands");
  String body;
  int code = 0;
  if (!doRequest("POST", url.c_str(), payload.c_str(), "application/json",
                 &body, &code, 256))
    return false;
  return code < 300;
}

bool Telegram::getUpdates(int64_t offset, int timeout_sec, TgUpdate *out,
                          int *count, int max_n, int64_t *next_offset) {
  if (count) *count = 0;
  if (next_offset) *next_offset = offset;
  String url = apiURL("getUpdates");
  url += "?timeout=";
  url += timeout_sec;
  url += "&limit=";
  url += max_n > 0 ? max_n : 1;
  if (offset > 0) {
    url += "&offset=";
    url += String((long)offset);
  }
  String body;
  int code = 0;
  if (!doRequest("GET", url.c_str(), nullptr, nullptr, &body, &code, 4096))
    return false;
  if (code >= 300) return false;
  if (body.indexOf("\"result\":[]") >= 0) return true;

  JsonDocument doc;
  DeserializationError err = deserializeJson(doc, body);
  if (err) {
    Serial.printf("tg json: %s\n", err.c_str());
    return false;
  }
  if (!doc["ok"].as<bool>()) return false;
  JsonArray result = doc["result"].as<JsonArray>();
  int n = 0;
  int64_t next = offset;
  for (JsonObject u : result) {
    if (n >= max_n) break;
    out[n].update_id = u["update_id"] | 0;
    out[n].has_message = false;
    out[n].text = "";
    out[n].chat_id = 0;
    if (u["message"].is<JsonObject>()) {
      JsonObject m = u["message"];
      out[n].has_message = true;
      out[n].chat_id = m["chat"]["id"] | (int64_t)0;
      const char *t = m["text"] | "";
      out[n].text = t;
    }
    if (out[n].update_id >= next) next = out[n].update_id + 1;
    n++;
  }
  if (count) *count = n;
  if (next_offset) *next_offset = next;
  return true;
}
